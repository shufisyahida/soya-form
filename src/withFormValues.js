import { set } from "lodash";
import React from "react";
import { connect } from "react-redux";
import { STATE_NAME } from "./_constants";
import createSoyaFormSelector from "./_selectors";
import { getDisplayName } from "./_utils";

export default (FORM_ID, fieldNames) => Component => {
  const mapStateToProps = state => {
    let formValues = {};
    if (state[STATE_NAME]) {
      const soyaFormSelector = createSoyaFormSelector(state, FORM_ID);
      formValues = fieldNames.reduce((formValues, fieldName) => {
        const fieldPath = Array.isArray(fieldName) ? fieldName : [fieldName];
        const value = soyaFormSelector.getFieldValue(fieldPath);
        set(formValues, fieldPath, value);
        return formValues;
      }, {});
    }
    return { formValues };
  };

  return connect(mapStateToProps)(
    class extends React.Component {
      static displayName = getDisplayName("WithFormValues", Component);

      render() {
        const { formValues, ...props } = this.props;
        delete props.dispatch;
        return <Component formValues={formValues} {...props} />;
      }
    }
  );
};
