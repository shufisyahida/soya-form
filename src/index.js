import createForm from './createForm';
import createField from './createField';
import createRepeatable from './createRepeatable';
import createSelector from './_selectors';
import withForm from './withForm';
import withFormValues from './withFormValues';

export {
  createForm,
  createField,
  createRepeatable,
  createSelector,
  withForm,
  withFormValues,
};
