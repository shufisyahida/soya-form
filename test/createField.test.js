import React from "react";
import { Provider } from "react-redux";
import TestRenderer from "react-test-renderer";
import { createConfigureStore } from "soya-next/redux";
import { createField, createForm, createSelector, withForm } from "../src";

let store;
const configureStore = createConfigureStore();

describe("createField", () => {
  beforeEach(() => {
    store = configureStore();
  });

  const noop = () => {}; // eslint-disable-line no-empty-function
  const InputField = createField(({ handleChange, name, value }) => (
    <input
      type="text"
      name={name}
      value={value}
      onChange={e => handleChange(e.target.value)}
    />
  ));
  const FORM_ID = "mockForm";
  const MockForm = createForm(FORM_ID)(({ children }) => (
    <form>{children}</form>
  ));

  it("should update form state on input change", () => {
    const renderer = TestRenderer.create(
      <Provider store={store}>
        <MockForm>
          <InputField name="foo" />
        </MockForm>
      </Provider>
    );
    const { root } = renderer;
    const foo = root.findByType("input");
    foo.props.onChange({ target: { value: "bar" } });
    const selector = createSelector(store.getState(), FORM_ID);
    expect(selector.getFieldValue(["foo"])).toBe("bar");
  });

  it("should call all change handlers on input change", () => {
    const changeHandlers = [jest.fn(), jest.fn()];
    const renderer = TestRenderer.create(
      <Provider store={store}>
        <MockForm>
          <InputField name="foo" changeHandlers={changeHandlers} />
        </MockForm>
      </Provider>
    );
    const { root } = renderer;
    const foo = root.findByType("input");
    foo.props.onChange({ target: { value: "bar" } });
    changeHandlers.forEach(handler => expect(handler).toBeCalledWith("bar"));
  });

  it("should validate on input change", () => {
    const changeValidators = [
      jest.fn(() => "Required"),
      jest.fn(() => "Min length is 5")
    ];
    const renderer = TestRenderer.create(
      <Provider store={store}>
        <MockForm>
          <InputField name="foo" changeValidators={changeValidators} />
        </MockForm>
      </Provider>
    );
    const { root } = renderer;
    const foo = root.findByType("input");
    foo.props.onChange({ target: { value: "bar" } });
    changeValidators.forEach(handler => expect(handler).toBeCalledWith("bar"));
    const selector = createSelector(store.getState(), FORM_ID);
    expect(selector.getField(["foo"])).toMatchSnapshot();
  });

  it("should lock the form on async validation", async () => {
    const asyncValidators = [
      jest.fn(() => Promise.resolve("Required")),
      jest.fn(() => Promise.resolve("Min length is 5"))
    ];
    const ButtonSave = withForm(({ form }) => (
      <button type="submit" onSubmit={async () => await form.submit(noop)}>
        Save
      </button>
    ));

    const renderer = TestRenderer.create(
      <Provider store={store}>
        <MockForm>
          <InputField name="foo" asyncValidators={asyncValidators} />
          <ButtonSave />
        </MockForm>
      </Provider>
    );
    const { root } = renderer;
    const foo = root.findByType("input");
    foo.props.onChange({ target: { value: "bar" } });
    const save = root.findByType("button");
    const submitPromise = save.props.onSubmit();
    let selector = createSelector(store.getState(), FORM_ID);
    expect(selector.getField(["foo"]).isValidating).toBe(true);
    await submitPromise;
    selector = createSelector(store.getState(), FORM_ID);
    expect(selector.getField(["foo"]).isValidating).toBe(false);
    asyncValidators.forEach(handler => expect(handler).toBeCalledWith("bar"));
  });
});
