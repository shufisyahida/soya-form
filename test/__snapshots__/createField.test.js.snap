// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createField should validate on input change 1`] = `
Object {
  "errorMessages": Array [
    "Required",
    "Min length is 5",
  ],
  "isEnabled": true,
  "isValidating": false,
  "touched": true,
  "value": "bar",
}
`;
